### PR TITLE
Feature/live processing

### PR DESCRIPTION
**Pull request by Vita1287**
**What is the purpose of this PR?**
This PR seeks to implement live processing so that the video previewer shows the correct time after each edit/delete, as requested by the export team.

**What was changed in this PR?**
liveProcess.ts created under editDelete. This file contains the functionality for live processing.
liveProcess.test.ts created under editDelete/__tests__. Tests are included for both deletion and editing.

### CODE REVIEW

File: src/main/editDelete/liveProcess.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * calculateTime calculates the outputStartTimes of a word based on the
+ * current state of all words in a Transcript. It takes into account edits/deletes
+ * @param word word in Transcript
+ * @param i the current index position in the words array
+ * @param words an array of words from Transcript
+ * @returns the input word with an updated outputStartTime
+ *
+ */
+
+const calculateTime = (word: Word, i: number, words: Word[]): Word => {
```
Comment (LiamTodd) [inline src/main/editDelete/liveProcess.ts]: Great stuff, would be good to add some comments in this function since the logic is quite complex and not super easy to grasp

**Review by LiamTodd — COMMENTED** [review]
Great work on the test cases - they're easy to understand and are all passing. Might be worth adding some more test cases for edge cases/more complicated series of operations on the transcript in future.

**Comment by Vita1287** [comment]
Yeah I could add some test cases combining edit and delete, just kinda ceebs

File: src/main/editDelete/liveProcess.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  return word;
+};
+
+/**
+ * Processes a Transcript so that all words have the correct outputStartTime
+ * @param transcript The original Transcript object
+ * @returns The Transcript object with updated outputStartTime for all words
+ */
+const liveProcessTranscript = (transcript: Transcription): Transcription => {
+  transcript.words.map(calculateTime);
```
Comment (chloebrett) [inline src/main/editDelete/liveProcess.ts]: Rather than mutating each word, how can you use `map` to produce a new transcript with the changed words?

File: src/main/editDelete/liveProcess.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * current state of all words in a Transcript. It takes into account edits/deletes
+ * @param word word in Transcript
+ * @param i the current index position in the words array
+ * @param words an array of words from Transcript
+ * @returns the input word with an updated outputStartTime
+ *
+ */
+
+const calculateTime = (word: Word, i: number, words: Word[]): Word => {
+  if (!word.deleted) {
```
Comment (chloebrett) [inline src/main/editDelete/liveProcess.ts]: If you flip this and do an early exit, the rest of the logic will be nested one level less deep which will make it easier to read

File: src/main/editDelete/liveProcess.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    if (i === 0) {
+      word.outputStartTime = 0;
+    } else {
+      let j = i - 1;
```
Comment (chloebrett) [inline src/main/editDelete/liveProcess.ts]: Like Liam said, a comment would be great here. Might be good to give j a descriptive name as well If you can.

File: src/main/editDelete/liveProcess.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      while (j > -1 && words[j].deleted) {
+        j -= 1;
+      }
+      if (j === -1) {
+        word.outputStartTime = 0;
+      } else {
+        word.outputStartTime = words[j].outputStartTime + words[j].duration;
```
Comment (chloebrett) [inline src/main/editDelete/liveProcess.ts]: Notice that you're only ever updating the outputStartTime, so how could you redesign this function to just return the outputStartTime you want?

**Review by chloebrett — COMMENTED** [review]
Looking good, just gave a couple of suggestions on how you might make this a little cleaner. Feel free to ask if you need further clarification

File: src/main/editDelete/liveProcess.ts (hunk 2 above)
Comment (v1ta1287) [reply src/main/editDelete/liveProcess.ts]: Good point. liveProcessTranscript now returns a new Transcript object.

File: src/main/editDelete/liveProcess.ts (hunk 3 above)
Comment (v1ta1287) [reply src/main/editDelete/liveProcess.ts]: Conditional statements definitely easier to read now with this change. Thanks

File: src/main/editDelete/liveProcess.ts (hunk 4 above)
Comment (v1ta1287) [reply src/main/editDelete/liveProcess.ts]: j has been renamed to nextNotDeleted, comments have been provided explaining the algorithm.

File: src/main/editDelete/liveProcess.ts (hunk 5 above)
Comment (v1ta1287) [reply src/main/editDelete/liveProcess.ts]: This is the only change request I'm not 100% sure about. I've refactored calculateTime to return the updated outputStartTime, but I now have to call another function processWord for the mapping. If there's any better suggestions please let me know and I'll try implement it.

File: src/main/editDelete/liveProcess.ts (hunk 1 above)
Comment (v1ta1287) [reply src/main/editDelete/liveProcess.ts]: Comments have been added to calculateTime

**Comment by v1ta1287** [comment]
Test case has been added which includes both edit and delete operations

**Review by LiamTodd — APPROVED** [review]

File: src/main/editDelete/liveProcess.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Processes a Transcript so that all words have the correct outputStartTime
+ * @param transcript The original Transcript object
+ * @returns The Transcript object with updated outputStartTime for all words
+ */
+const liveProcessTranscript = (transcript: Transcription): Transcription => {
+  return {
+    confidence: transcript.confidence,
+    words: transcript.words.map(processWord),
```
Comment (chloebrett) [inline src/main/editDelete/liveProcess.ts:59]: Nice work! You can remove the need for `processWord` like this:

`words: transcript.words.map((word, i, words) => ({ ...word, outputStartTime: calculateTime(word, i, words), })),`

Or make `return { ...word, outputStartTime: calculateTime(word, i, words), }` the body of processWord, but with proper line breaks probably not necessary

**Review by chloebrett — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by rileykeane — APPROVED** [review]
Looks great!
